### PR TITLE
Hashing strategies v2 - Closes #1483

### LINKED ISSUE #1483: Determining cache misses (or hits) is slow for jobs with bam file inputs on a local backend
- local backend

The slowness is probably driven by calculating a md5 hash on the bam input files.  For jobs that do not have bam files as inputs, the cache determination is fast.

Calculating a md5 on a bam file is too slow for use in call caching on a local backend.  I can provide a time estimate, if necessary.

Does SGE backend has the same issue?

Proposed solution:
- If md5 file is next to input file (e.g. sample1.bam is next to sample1.bam.md5) then read the md5 hash from the *.md5 file.  Otherwise, use the file path.  For bams, the file paths will tend to be static for most local backend environments anyway.  
- Make sure this convention is documented.
- Perhaps have a flag in the conf file so that users can choose which convention they prefer?  IMHO, have the fast method (above) as the default.

### PR DESCRIPTION
**Pull request by Horneth**

### CODE REVIEW

**Comment by mcovarr** [comment]
:+1:

File: supportedBackends/sfs/src/main/scala/cromwell/backend/impl/sfs/config/ConfigHashingStrategy.scala [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        case "path" => HashPathStrategy(checkSiblingMd5)
+        case "file" => HashFileStrategy(checkSiblingMd5)
+        case what =>
+          logger.warn(s"Unrecognized hashing strategy $what.")
+          HashPathStrategy(checkSiblingMd5)
+      }
+  }
+}
+
+abstract class ConfigHashingStrategy {
```
Comment (geoffjentry) [inline supportedBackends/sfs/src/main/scala/cromwell/backend/impl/sfs/config/ConfigHashingStrategy.scala]: can this be sealed?

File: supportedBackends/sfs/src/main/scala/cromwell/backend/sfs/BackgroundAsyncJobExecutionActor.scala [hunk 2]
```diff
@@ -13,10 +13,10 @@ trait BackgroundAsyncJobExecutionActor extends SharedFileSystemAsyncJobExecution
     val stdout = pathPlusSuffix(jobPaths.stdout, "background")
     val stderr = pathPlusSuffix(jobPaths.stderr, "background")
     val argv = Seq("/bin/bash", backgroundScript)
-    new ProcessRunner(argv, stdout, stderr)
+    new ProcessRunner(argv, stdout.path, stderr.path)
   }
 
-  private def writeBackgroundScript(backgroundScript: Path, backgroundCommand: String): Unit = {
+  private def writeBackgroundScript(backgroundScript: File, backgroundCommand: String): Unit = {
```
Comment (geoffjentry) [inline supportedBackends/sfs/src/main/scala/cromwell/backend/sfs/BackgroundAsyncJobExecutionActor.scala]: Why the change from path to file?

File: core/src/main/scala/cromwell/core/callcaching/CallCachingHashingStrategy.scala [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package cromwell.core.callcaching
+
+sealed trait CallCachingHashingStrategy
+
+case class ExistingSiblingHash(extension: String) extends CallCachingHashingStrategy
+case object PathHash extends CallCachingHashingStrategy
+case object ComputeHash extends CallCachingHashingStrategy
```
Comment (cjllanwarne) [inline core/src/main/scala/cromwell/core/callcaching/CallCachingHashingStrategy.scala]: On behalf of @geoffjentry, `final`ise these

File: core/src/main/scala/cromwell/util/FileUtil.scala [hunk 4]
```diff
@@ -23,4 +24,16 @@ object FileUtil {
...
+
+  @tailrec
+  final private def followSymlinks(file: better.files.File): better.files.File = {
+    file.symbolicLink match {
+      case Some(target) => followSymlinks(target)
+      case None => file
+    }
+  }
+
+  implicit class EnhancedBetterFile(val file: better.files.File) extends AnyVal {
```
Comment (cjllanwarne) [inline core/src/main/scala/cromwell/util/FileUtil.scala]: Not `EvenBetterFile`? 😢

File: supportedBackends/sfs/src/main/scala/cromwell/backend/impl/sfs/config/ConfigBackendLifecycleActorFactory.scala [hunk 5]
```diff
@@ -32,6 +38,10 @@ class ConfigBackendLifecycleActorFactory(val configurationDescriptor: BackendCon
     initializationData.runtimeAttributesBuilder.definitions.toSet
   }
 
-  override lazy val fileHashingFunction: Option[FileHashingFunction] = Option(FileHashingFunction(ConfigBackendFileHashing.getMd5Result))
+  override lazy val fileHashingFunction: Option[FileHashingFunction] = {
+    logger.info(hashingStrategy.toString)
```
Comment (cjllanwarne) [inline supportedBackends/sfs/src/main/scala/cromwell/backend/impl/sfs/config/ConfigBackendLifecycleActorFactory.scala]: I would 
- Add some sort of context to this hashingStrategy string message
- Make it debug

File: supportedBackends/sfs/src/main/scala/cromwell/backend/impl/sfs/config/ConfigHashingStrategy.scala [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+case class HashPathStrategy(checkSiblingMd5: Boolean) extends ConfigHashingStrategy {
+  override def hash(file: File): Try[String] = {
+    Try(DigestUtils.md5Hex(file.path.toAbsolutePath.toString))
+  }
+
+  override val description = "hash file path"
+}
+case class HashFileStrategy(checkSiblingMd5: Boolean) extends ConfigHashingStrategy {
```
Comment (cjllanwarne) [inline supportedBackends/sfs/src/main/scala/cromwell/backend/impl/sfs/config/ConfigHashingStrategy.scala]: `final`s here too

File: supportedBackends/sfs/src/main/scala/cromwell/backend/impl/sfs/config/ConfigHashingStrategy.scala (hunk 1 above)
Comment (cjllanwarne) [inline supportedBackends/sfs/src/main/scala/cromwell/backend/impl/sfs/config/ConfigHashingStrategy.scala]: But what if we want to make a java interop? 😛

File: supportedBackends/sfs/src/main/scala/cromwell/backend/sfs/SharedFileSystem.scala [hunk 7]
```diff
@@ -29,51 +29,44 @@ object SharedFileSystem {
     }
   }
 
-  type PathsPair = (Path, Path)
-  type DuplicationStrategy = (Path, Path) => Try[Unit]
+  type FilesPair = (File, File)
+  type DuplicationStrategy = (File, File) => Try[Unit]
```
Comment (cjllanwarne) [inline supportedBackends/sfs/src/main/scala/cromwell/backend/sfs/SharedFileSystem.scala]: I would really like to ADT these :)

At the very least it should be `DuplicationStrategy = FilesPair => Try[Unit]`

File: supportedBackends/sfs/src/main/scala/cromwell/backend/sfs/SharedFileSystem.scala [hunk 8]
```diff
@@ -29,51 +29,44 @@ object SharedFileSystem {
...
 
-  /**
+    /**
```
Comment (cjllanwarne) [inline supportedBackends/sfs/src/main/scala/cromwell/backend/sfs/SharedFileSystem.scala]: I think this was right before. You want the `*`s to align

File: supportedBackends/sfs/src/main/scala/cromwell/backend/sfs/SharedFileSystem.scala [hunk 9]
```diff
@@ -29,51 +29,44 @@ object SharedFileSystem {
...
-
-  private def localizePathViaSymbolicLink(originalPath: Path, executionPath: Path): Try[Unit] = {
-      if (File(originalPath).isDirectory) Failure(new UnsupportedOperationException("Cannot localize directory with symbolic links"))
+  private def localizePathViaSymbolicLink(originalPath: File, executionPath: File): Try[Unit] = {
+      if (originalPath.isDirectory) Failure(new UnsupportedOperationException("Cannot localize directory with symbolic links"))
       else {
-        File(executionPath).parent.createDirectories()
-        Try(Files.createSymbolicLink(executionPath, originalPath.toAbsolutePath))
+        executionPath.parent.createDirectories()
+        Try(executionPath.linkTo(originalPath, symbolic = true)) map { _ => executionPath }
```
Comment (cjllanwarne) [inline supportedBackends/sfs/src/main/scala/cromwell/backend/sfs/SharedFileSystem.scala]: Could you comment this? It's a little bit funky?

`"Try to create a symlink (which we ignore), and if that works return the original, otherwise return the linking failure"`?

File: supportedBackends/sfs/src/test/scala/cromwell/backend/impl/sfs/config/ConfigHashingStrategySpec.scala [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+import wdl4s.values.WdlFile
+
+import scala.util.Success
+
+class ConfigHashingStrategySpec extends FlatSpec with Matchers with TableDrivenPropertyChecks with Mockito with BeforeAndAfterAll {
+
+  behavior of "ConfigHashingStrategy"
+
+  val steak = "Steak"
+  val steakHash = DigestUtils.md5Hex(steak)
```
Comment (cjllanwarne) [inline supportedBackends/sfs/src/test/scala/cromwell/backend/impl/sfs/config/ConfigHashingStrategySpec.scala]: Steak hash sounds delicious 😋

File: supportedBackends/sfs/src/main/scala/cromwell/backend/sfs/SharedFileSystem.scala (hunk 7 above)
Comment (Horneth) [inline supportedBackends/sfs/src/main/scala/cromwell/backend/sfs/SharedFileSystem.scala]: I agree for the pair but for the strategy I think it's really easier to manipulate function `type`s than function `class`es

File: supportedBackends/sfs/src/main/scala/cromwell/backend/sfs/SharedFileSystem.scala (hunk 9 above)
Comment (Horneth) [inline supportedBackends/sfs/src/main/scala/cromwell/backend/sfs/SharedFileSystem.scala]: `executionPath.linkTo(originalPath)` returns `originalPath`, which I agree is counter intuitive but it's what it is. However here we want to return the executionPath (the actual link), not the target.
I'll add a comment.

File: supportedBackends/sfs/src/main/scala/cromwell/backend/impl/sfs/config/ConfigBackendLifecycleActorFactory.scala (hunk 5 above)
Comment (Horneth) [inline supportedBackends/sfs/src/main/scala/cromwell/backend/impl/sfs/config/ConfigBackendLifecycleActorFactory.scala]: The message is explicit (https://github.com/broadinstitute/cromwell/pull/1508/files#diff-1a39f56f5db2136bb3d1e898de8671a5R56)
I'll change it to debug

**Comment by cjllanwarne** [comment]
👍
